lib/hash_ring/ets.ex

```diff
@@ -41,202 +59,233 @@ defmodule ExHashRing.HashRing.ETS do
     GenServer.start_link(__MODULE__, {name, nodes, num_replicas, overrides}, gen_opts)
   end
 
-  @spec init({atom, [binary], integer, override_map}) :: {:ok, t}
-  def init({name, nodes, default_num_replicas, overrides}) do
-    table =
-      :ets.new(:ring, [
-        :protected,
-        :ordered_set,
-        {:read_concurrency, true}
-      ])
-
-    state =
-      rebuild(%__MODULE__{
-        table: table,
-        default_num_replicas: default_num_replicas,
-        nodes: transform_nodes(nodes, default_num_replicas),
-        overrides: overrides,
-        name: name
-      })
-
-    {:ok, state}
-  end
-
-  def stop(name) do
-    GenServer.stop(name)
-  end
-
-  @spec set_nodes(atom, [binary | {binary, integer}]) :: {:ok, [{binary, integer}]}
-  def set_nodes(name, node_names) do
-    GenServer.call(name, {:set_nodes, node_names})
-  end
-
+  @doc """
+  Adds a node to the existing set of nodes in the ring.
+  """
   @spec add_node(atom, binary, integer) :: {:ok, [{binary, integer}]} | {:error, :node_exists}
   def add_node(name, node_name, num_replicas \\ nil) do
     GenServer.call(name, {:add_node, node_name, num_replicas})
   end
 
-  @spec remove_node(atom, binary) :: {:ok, [{binary, integer}]} | {:error, :node_not_exists}
-  def remove_node(name, node_name) do
-    GenServer.call(name, {:remove_node, node_name})
+  @doc """
+  Adds multiple nodes to the existing set of nodes in the ring.
+  """
+  @spec add_nodes(name :: atom, nodes :: [binary | {binary, integer}]) :: {:ok, [{binary, integer}]} | {:error, :node_exists}
+  def add_nodes(name, nodes) do
+    GenServer.call(name, {:add_nodes, nodes})
   end
 
-  @spec set_overrides(atom, override_map) :: {:ok, override_map}
-  def set_overrides(name, overrides) do
-    GenServer.call(name, {:set_overrides, overrides})
+  @doc """
+  Finds the node responsible for the given key in the specified ring.
+  """
+  @spec find_node(atom, binary | integer) :: {:ok, binary} | {:error, atom}
+  def find_node(name, key) do
+    case Config.get(name) do
+      {:ok, {{table, num_nodes}, _previous_table, gen, overrides}} when num_nodes > 0 ->
+        case overrides do
+          %{^key => overrides} ->
+            {:ok, hd(overrides)}
```

Comment:
if hd is empty shouldn't we log error and at least try and use the normal ring?

Reply:
The override list is dropped if it's empty to prevent this in `update_overrides/2`.

See https://github.com/discord/ex_hash_ring/pull/10/files/b34ca57db26b165b73cb231f024996f8e75ce4fb#diff-8495c75f3d229f289b88e4080201c480R532-R540 for details.